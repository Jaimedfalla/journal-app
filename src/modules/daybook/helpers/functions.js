import axios from "axios";

const months = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio','Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']
const days   = ['Domingo','Lunes','Martes','Miércoles','Jueves','Viernes','Sábado']

const getDayMonthYear =(dateString)=>{
    const date = new Date(dateString);

    return{
        day:date.getDate(),
        month:months[date.getMonth()],
        yearDay:`${date.getFullYear()}, ${days[date.getDay()]}`
    }
}

const uploadImage = async (file)=>{
    if(!file) return

    try{
        const formData = new FormData()
        formData.append('upload_preset','dailybook')
        formData.append('file',file)

        const url = 'https://api.cloudinary.com/v1_1/dstyfk698/image/upload'
        const {data} = await axios.post(url,formData)

        return data.secure_url;

    }catch(error){
        console.error('Error al cargar la imagen, revisar logs')
        console.log(error)
        return null
    }
}

export {getDayMonthYear,uploadImage};