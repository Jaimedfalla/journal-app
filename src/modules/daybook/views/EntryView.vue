<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-sucess fs-3 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-3">{{ month }}</span>
        <span class="mx-2 fs-4 f2-light">{{ yearDay }}</span>
      </div>

      <div>
        <input v-show="false"
          type="file"
          @change="onSelectedImage"
          ref="imageSelector"
          accept="image/png, image/jpeg, image/jpg">

        <button v-if="entry.id" @click="onDeleteEntry"
          class="btn btn-danger mx-2">
            Borrar
          <i class="fa fa-trash-alt"></i>
        </button>
        <button
          @click="selectImage"
          class="btn btn-primary">
          Subir foto
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>
    <hr />
    <div class="d-flex flex-column px-3 h-75">
      <textarea placeholder="¿Qué sucedió hoy?" v-model="entry.text"></textarea>
    </div>
    <img v-if="entry.picture && !localImage"
      :src="entry.picture"
      alt="entry-picture"
      class="img-thumbnail"
    />
    <img v-if="localImage"
      :src="localImage"
      alt="entry-picture"
      class="img-thumbnail"
    />
  </template>

  <Fab :icon="'fa-save'"
    @on:click="saveEntry"/>

</template>

<script>
import { defineAsyncComponent } from "vue";
import { mapGetters, mapActions } from "vuex";
import {getDayMonthYear,uploadImage} from "../helpers/functions";
import Swal from 'sweetalert2'

export default {
  name:'EntryView',
  components: {
    Fab: defineAsyncComponent(() =>
      import("../components/FloatingActionButton.vue")
    ),
  },
  data() {
    return {
      entry: {
        text:'',
        date: new Date().toISOString()
      },
      save:this.createEntry,
      localImage:null,
      file:null
    };
  },
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  computed: {
    ...mapGetters("journal", {
      getById: "getEntryById",
    }),
    day() {
      const { day } = getDayMonthYear(this.entry.date);

      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);

      return month;
    },
    yearDay() {
      const { yearDay } = getDayMonthYear(this.entry.date);

      return yearDay;
    },
  },
  methods: {
    ...mapActions('journal',['updateEntry','createEntry','deleteEntry']),
    loadEntry() {
      let entry;
      if(this.id ==='new'){
        this.entry = {
          text:'',
          date: new Date().toISOString()
        }
        this.save = this.createEntry
        this.localImage = null
        return
      }
      
      entry = this.getById(this.id);
      if (!entry) return this.$router.push({ name: "no-entry" });

      this.save = this.updateEntry
      this.entry = entry;
    },
    onSelectedImage(event){
      const file = event.target.files[0]
      if(!file){
        this.localImage = null
        this.file = null
        return
      }

      this.file=file
      const fr = new FileReader()
      fr.onload = ()=> this.localImage = fr.result
      fr.readAsDataURL(file)

    },
    selectImage(){
      this.$refs.imageSelector.click()
    },
    async saveEntry(){
      return new Promise(console.log('Guardando entrada'))
      // const creatEntry = async ()=>{
      //     const id = await this.save(this.entry)||this.id;
      //     this.$router.push({ name: "entry",params:{id} });
      //   }

      // const msgs = {
      //     title:'Guardado',
      //     text: 'Entrada registrada con éxito'
      //   }
      
      // console.log(creatEntry)
      // this.showAlerts(creatEntry,msgs)
    },
    async onDeleteEntry(){
      const {isConfirmed} = await Swal.fire({
        title:'¿Está seguro',
        text:'Una vez borrado, no se podrá recuperar',
        showDenyButton:true,
        confirmButtonText:'Sí, estoy seguro'
      })

      if(isConfirmed){
        this.showAlerts(async ()=>{
          await this.deleteEntry(this.id)
          this.$router.push({name:'no-entry'})
        },{
          title:'Eliminado',
          text:''
        })
      }
    },
    async showAlerts(method,message){
      Swal.fire({
          title:'Espere por favor',
          allowOutsideClick:false
        })

      Swal.showLoading()
      const picture = await uploadImage(this.file)
      this.entry.picture = picture;
      await method()
      Swal.fire(message.title,message.text,'success')
      this.file = null
      this.localImage = null
    }
  },
  watch: {
    id() {
      this.loadEntry();
    },
  },
  created() {
    this.loadEntry();
  },
};
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0px 5px 10px rgba($color: black, $alpha: 0.2);
}
</style>