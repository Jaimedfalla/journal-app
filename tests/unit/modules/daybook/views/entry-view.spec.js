import { journalState } from "../__mocks__/test-journal-state"
import {createVuexStore} from '../__mocks__/global-functions'
import { shallowMount } from "@vue/test-utils"
import EntryView from "@/modules/daybook/views/EntryView.vue"
import Swal from "sweetalert2"
import * as functions from '@/modules/daybook/helpers/functions'

//Para emular el comportamiento de SweetAlert
jest.mock('sweetalert2',()=>({
    fire:jest.fn(),
    showLoading:jest.fn()
}))

describe('Pruebas en el entry view',()=>{

    const store = createVuexStore(journalState)
    store.dispatch = jest.fn()

    const mockRouter = {
        push:jest.fn()
    }

    let wrapper

    beforeEach(()=>{
        jest.clearAllMocks()

        wrapper = shallowMount(EntryView,{
            props:{
                id:'-NLMbSJH-FO9UimMeioF'
            },
            global:{
                mocks:{
                    $router:mockRouter
                },
                plugins:[store]
            },
        })
    })

    it('Debe sacar al usuario porque el id no existe',()=>{
        shallowMount(EntryView,{
            props:{
                id:'Este id no existe en el store'
            },
            global:{
                mocks:{
                    $router:mockRouter
                },
                plugins:[store]
            },
        })

        expect(mockRouter.push).toHaveBeenCalledWith({name:'no-entry'})
    })

    it('Debe mostrar la entrada correctamente',()=>{
        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled()
    })

    it('Debe borrar la entrada y salir',(done)=>{
        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed:true}))
        wrapper.find('.btn-danger').trigger('click')
        expect(Swal.fire).toHaveBeenCalledWith({
            title:'¿Está seguro',
            text:'Una vez borrado, no se podrá recuperar',
            showDenyButton:true,
            confirmButtonText:'Sí, estoy seguro'
        })

        setTimeout(()=>{
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry','-NLMbSJH-FO9UimMeioF')
            expect(mockRouter.push).toHaveBeenCalled()
            done()
        },1)
    })

    it('Debe guardar la nueva entrada',async()=>{
        const wrapper = shallowMount(EntryView,{
            props:{
                id:'new'
            },
            global:{
                mocks:{
                    $router:mockRouter
                },
                plugins:[store]
            },
        })
 
        const uploadSpy = jest.spyOn(functions,'uploadImage')
            .mockResolvedValueOnce(Promise.resolve("https://www.google.com"))
            
        const textarea = wrapper.find('textarea')
        textarea.text = 'Entrada del Test unitario'
        await wrapper.find('fab-stub').trigger('click')        
        expect(uploadSpy).toHaveBeenCalledTimes(1)
        expect(Swal.fire).toHaveBeenCalledWith({
            title:'Espere por favor',
            allowOutsideClick:false
          })
    })
})