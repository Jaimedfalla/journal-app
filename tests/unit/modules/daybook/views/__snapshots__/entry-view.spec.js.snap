// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el entry view Debe mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-sucess fs-3 fw-bold">9</span><span class="mx-1 fs-3">Enero</span><span class="mx-2 fs-4 f2-light">2023, Lunes</span></div>
  <div><input type="file" accept="image/png, image/jpeg, image/jpg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedió hoy?"></textarea></div>
<img src="https://res.cloudinary.com/dstyfk698/image/upload/v1673285980/dailybook/sdgkulkge2nm9dvnkoop.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
